File: examples/companion-packages/dotnet-fsharp/test/AgentXM.Example.TinyFlags.FSharp.Tests/TinyFlagsTests.fs
module AgentXM.Example.TinyFlags.FSharp.Tests.TinyFlagsTests

open System
open Expecto
open AgentXM.Example.TinyFlags.FSharp

let private context =
    { EvaluationContext.empty with UserId = Some "user-1" }

[<Tests>]
let tests =
    testList "TinyFlags" [
        testCase "boolean flags use defaults when no rollout is configured" <| fun _ ->
            let flags =
                TinyFlags.create [
                    "checkoutRedesign", Flag.Boolean(defaultValue = true)
                ]

            Expect.isTrue
                (flags |> TinyFlags.enabled "checkoutRedesign" context)
                "default value should be returned"

        testCase "boolean rollout boundaries are deterministic" <| fun _ ->
            let flags =
                TinyFlags.create [
                    "disabledExperiment", Flag.Boolean(defaultValue = false, rollout = 0)
                    "enabledExperiment", Flag.Boolean(defaultValue = false, rollout = 100)
                ]

            Expect.isFalse
                (flags |> TinyFlags.enabled "disabledExperiment" context)
                "0% rollout should be disabled"
            Expect.isTrue
                (flags |> TinyFlags.enabled "enabledExperiment" context)
                "100% rollout should be enabled"
            Expect.equal
                (flags |> TinyFlags.enabled "enabledExperiment" context)
                (flags |> TinyFlags.enabled "enabledExperiment" context)
                "repeated evaluations should be deterministic"

        testCase "variant flags return the default outside rollout allocations" <| fun _ ->
            let flags =
                TinyFlags.create [
                    "searchRanking",
                    Flag.Variant(
                        variants = [ "classic"; "semantic" ],
                        defaultValue = "classic",
                        rollout = Map.ofList [ "semantic", 0 ]
                    )
                ]

            Expect.equal
                (flags |> TinyFlags.variant "searchRanking" context)
                "classic"
                "no allocation should fall back to default"

        testCase "variant flags can allocate all traffic to a variant" <| fun _ ->
            let flags =
                TinyFlags.create [
                    "searchRanking",
                    Flag.Variant(
                        variants = [ "classic"; "semantic" ],
                        defaultValue = "classic",
                        rollout = Map.ofList [ "semantic", 100 ]
                    )
                ]

            Expect.equal
                (flags |> TinyFlags.variant "searchRanking" context)
                "semantic"
                "100% allocation should win"

        testCase "evaluate returns a typed FlagValue for each kind" <| fun _ ->
            let flags =
                TinyFlags.create [
                    "checkoutRedesign", Flag.Boolean(defaultValue = true)
                    "searchRanking",
                    Flag.Variant(
                        variants = [ "classic"; "semantic" ],
                        rollout = Map.ofList [ "semantic", 100 ]
                    )
                ]

            Expect.equal
                (flags |> TinyFlags.evaluate "checkoutRedesign" context)
                (BoolValue true)
                "boolean flag should evaluate to BoolValue"
            Expect.equal
                (flags |> TinyFlags.evaluate "searchRanking" context)
                (VariantValue "semantic")
                "variant flag should evaluate to VariantValue"

        testCase "boolean rollout above 100 fails at construction time" <| fun _ ->
            Expect.throwsT<ArgumentException>
                (fun () -> Flag.Boolean(defaultValue = false, rollout = 101) |> ignore)
                "rollout above 100 should throw"

        testCase "variant default must be one of the variants" <| fun _ ->
            Expect.throwsT<ArgumentException>
                (fun () ->
                    Flag.Variant(variants = [ "classic"; "semantic" ], defaultValue = "personalized")
                    |> ignore)
                "default outside variants should throw"

        testCase "variant rollout totals above 100 fail at construction time" <| fun _ ->
            Expect.throwsT<ArgumentException>
                (fun () ->
                    Flag.Variant(
                        variants = [ "classic"; "semantic" ],
                        defaultValue = "classic",
                        rollout = Map.ofList [ "semantic", 80; "classic", 30 ]
                    )
                    |> ignore)
                "rollout totals above 100 should throw"
    ]

File: examples/companion-packages/dotnet-csharp-nuget-lib/test/AgentXM.Examples.TinyFlags.CSharp.Tests/TinyFlagsTests.cs
using Xunit;

namespace AgentXM.Examples.TinyFlags.CSharp.Tests;

public class TinyFlagsTests
{
    [Fact]
    public void BooleanFlagsUseDefaults()
    {
        var flags = TinyFlags.Create(new Dictionary<string, FlagDefinition>
        {
            ["checkoutRedesign"] = TinyFlag.Boolean(defaultValue: true),
        });

        Assert.True(flags.Enabled("checkoutRedesign", new EvaluationContext(UserId: "user-1")));
    }

    [Fact]
    public void BooleanRolloutBoundariesAreDeterministic()
    {
        var flags = TinyFlags.Create(new Dictionary<string, FlagDefinition>
        {
            ["disabledExperiment"] = TinyFlag.Boolean(defaultValue: false, rollout: 0),
            ["enabledExperiment"] = TinyFlag.Boolean(defaultValue: false, rollout: 100),
        });
        var ctx = new EvaluationContext(UserId: "user-1");

        Assert.False(flags.Enabled("disabledExperiment", ctx));
        Assert.True(flags.Enabled("enabledExperiment", ctx));
        Assert.Equal(flags.Enabled("enabledExperiment", ctx), flags.Enabled("enabledExperiment", ctx));
    }

    [Fact]
    public void VariantFlagsReturnDefaultsOutsideRolloutAllocations()
    {
        var flags = TinyFlags.Create(new Dictionary<string, FlagDefinition>
        {
            ["searchRanking"] = TinyFlag.Variant(
                ["classic", "semantic"],
                defaultValue: "classic",
                rollout: new Dictionary<string, int> { ["semantic"] = 0 }),
        });

        Assert.Equal("classic", flags.Variant("searchRanking", new EvaluationContext(UserId: "user-1")));
    }

    [Fact]
    public void VariantFlagsCanAllocateAllTrafficToAVariant()
    {
        var flags = TinyFlags.Create(new Dictionary<string, FlagDefinition>
        {
            ["searchRanking"] = TinyFlag.Variant(
                ["classic", "semantic"],
                defaultValue: "classic",
                rollout: new Dictionary<string, int> { ["semantic"] = 100 }),
        });

        Assert.Equal("semantic", flags.Variant("searchRanking", new EvaluationContext(UserId: "user-1")));
    }

    [Fact]
    public void VariantRolloutAllocatesInDeclaredOrder()
    {
        // Each variant's bucket allocation depends on the order it was declared
        // in, not on the order of the rollout dictionary.
        var flags = TinyFlags.Create(new Dictionary<string, FlagDefinition>
        {
            ["split"] = TinyFlag.Variant(
                ["first", "second"],
                defaultValue: "first",
                rollout: new Dictionary<string, int> { ["second"] = 50, ["first"] = 50 }),
        });

        for (var i = 0; i < 50; i++)
        {
            var resolved = flags.Variant("split", new EvaluationContext(UserId: $"user-{i}"));
            Assert.True(resolved is "first" or "second");
        }
    }

    [Fact]
    public void InvalidBooleanRolloutThrows()
    {
        Assert.Throws<ArgumentOutOfRangeException>(() => _ = TinyFlag.Boolean(rollout: 101));
    }

    [Fact]
    public void InvalidVariantDefaultThrows()
    {
        Assert.Throws<ArgumentException>(() => _ = TinyFlag.Variant(["classic", "semantic"], defaultValue: "personalized"));
    }

    [Fact]
    public void VariantRolloutExceedingHundredThrows()
    {
        Assert.Throws<ArgumentOutOfRangeException>(() => _ = TinyFlag.Variant(
            ["classic", "semantic"],
            rollout: new Dictionary<string, int> { ["semantic"] = 80, ["classic"] = 30 }));
    }

    [Fact]
    public void EvaluateReturnsTypedResultForEachFlagKind()
    {
        var flags = TinyFlags.Create(new Dictionary<string, FlagDefinition>
        {
            ["onboarding"] = TinyFlag.Boolean(defaultValue: true),
            ["searchRanking"] = TinyFlag.Variant(
                ["classic", "semantic"],
                defaultValue: "semantic",
                rollout: new Dictionary<string, int> { ["semantic"] = 100 }),
        });
        var ctx = new EvaluationContext(UserId: "user-1");

        Assert.Equal(new FlagValue.Bool(true), flags.Evaluate("onboarding", ctx));
        Assert.Equal(new FlagValue.Variant("semantic"), flags.Evaluate("searchRanking", ctx));
    }

    [Fact]
    public void FlagValueExposesTryGetAccessors()
    {
        FlagValue boolValue = new FlagValue.Bool(true);
        FlagValue variantValue = new FlagValue.Variant("semantic");

        Assert.True(boolValue.TryGetBool(out var b));
        Assert.True(b);
        Assert.False(boolValue.TryGetVariant(out _));

        Assert.True(variantValue.TryGetVariant(out var v));
        Assert.Equal("semantic", v);
        Assert.False(variantValue.TryGetBool(out _));
    }

    [Fact]
    public void ParamsOverloadBuildsVariantFromInlineValues()
    {
        var flag = TinyFlag.Variant("classic", "semantic");

        Assert.Equal(2, flag.Variants.Length);
        Assert.Equal("classic", flag.Variants[0]);
        Assert.Equal("semantic", flag.Variants[1]);
        Assert.Equal("classic", flag.Default);
    }
}
